telethon/utils.py

```diff
@@ -601,14 +601,15 @@ def get_message_id(message):
 
 
 def _get_metadata(file):
-    # `hachoir` only deals with paths to in-disk files, while
-    # `_get_extension` supports a few other things. The parser
-    # may also fail in any case and we don't want to crash if
+    # The parser may fail and we don't want to crash if
     # the extraction process fails.
-    if hachoir and isinstance(file, str) and os.path.isfile(file):
+    if hachoir:
         try:
-            with hachoir.parser.createParser(file) as parser:
-                return hachoir.metadata.extractMetadata(parser)
+            file = io.BytesIO(file) if isinstance(file, bytes) else file
+            parser = hachoir.parser.createParser(file)
+            if isinstance(file, str):
+                parser.close()
+            return hachoir.metadata.extractMetadata(parser)
         except Exception as e:
             _log.warning('Failed to analyze %s: %s %s', file, e.__class__, e)
```

Comment:
This will now log a warning when trying to use file streams. Do you think it would be better to gracefully handle out of date hachoir versions?

Reply:
I think that's a good idea. I made a commit to reflect what you suggested, but I think I made a bit of a mess of the commits. What do you think I should do?

Reply:
Maybe checking for `hachoir.__version__` would be better, since the `AssertionError` could come from elsewhere.

Reply:
How do you suggest we check the version? There are a few ways (all of the following are available in Python 3.5.0 and higher):

1. There is `pkg_resources.parse_version()` from `setuptools` which works, but since `setuptools` is not guaranteed to be installed, I'm not sure about it.
2. There is also `distutils.version` which is old, but still works.
3. It's unreliable but we could do a simple `int(hachoir.__version__.replace('.','')) < 312`

I guess we could do something like this:
```python
try:
    from pkg_resources import parse_version
except ImportError:
    from distutils.version import StrictVersion as parse_version
```
and then check:
```python
def _get_metadata(file):
    # The parser may fail and we don't want to crash if
    # the extraction process fails.
    if hachoir:
        try:
            file = io.BytesIO(file) if isinstance(file, bytes) else file
            parser = hachoir.parser.createParser(file)
            if isinstance(file, str):
                parser.close()
            return hachoir.metadata.extractMetadata(parser)
        except Exception as e:
            if parse_version(hachoir.__version__) < parse_version('3.1.2'):
                msg = ('Your version of Hachoir can only open in-disk files.'
                       ' Update it to 3.1.2 or higher to read byte arrays/streams.')
                _log.warning('Failed to analyze %s: %s', file, msg)
            else:
                _log.warning('Failed to analyze %s: %s %s', file, e.__class__, e)
```